presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionFactory.java

```diff
@@ -0,0 +1,291 @@
+
+        String user = trimEmptyToNull(servletRequest.getHeader(PRESTO_USER));
+        assertRequest(user != null, "User must be set");
+        identity = new Identity(user, Optional.ofNullable(servletRequest.getUserPrincipal()));
+
+        source = servletRequest.getHeader(PRESTO_SOURCE);
+        userAgent = servletRequest.getHeader(USER_AGENT);
+        remoteUserAddress = servletRequest.getRemoteAddr();
+        timeZoneId = servletRequest.getHeader(PRESTO_TIME_ZONE);
+        language = servletRequest.getHeader(PRESTO_LANGUAGE);
+
+        // parse session properties
+        for (Entry<String, String> entry : parseSessionHeaders(servletRequest).entrySet()) {
+            String fullPropertyName = entry.getKey();
+            String propertyValue = entry.getValue();
+            List<String> nameParts = Splitter.on('.').splitToList(fullPropertyName);
+            if (nameParts.size() == 1) {
+                String propertyName = nameParts.get(0);
+
+                assertRequest(!propertyName.isEmpty(), "Invalid %s header", PRESTO_SESSION);
+
+                // catalog session properties can not be validated until the transaction has stated, so we delay system property validation also
+                systemProperties.put(propertyName, propertyValue);
+            }
+            else if (nameParts.size() == 2) {
+                String catalogName = nameParts.get(0);
+                String propertyName = nameParts.get(1);
+
+                assertRequest(!catalogName.isEmpty(), "Invalid %s header", PRESTO_SESSION);
+                assertRequest(!propertyName.isEmpty(), "Invalid %s header", PRESTO_SESSION);
+
+                // catalog session properties can not be validated until the transaction has stated
+                catalogSessionProperties.computeIfAbsent(catalogName, id -> new HashMap<>()).put(propertyName, propertyValue);
+            }
+            else {
+                throw badRequest(format("Invalid %s header", PRESTO_SESSION));
+            }
+        }
+
+        preparedStatements = parsePreparedStatementsHeaders(servletRequest);
+
+        String transactionIdHeader = servletRequest.getHeader(PRESTO_TRANSACTION_ID);
+        clientTransactionSupport = transactionIdHeader != null;
+        transactionId = parseTransactionId(transactionIdHeader);
+    }
+
+    @Override
+    public Identity getIdentity()
+    {
+        return identity;
+    }
+
+    @Override
+    public Session createSession(
+            QueryId queryId,
+            TransactionManager transactionManager,
+            AccessControl accessControl,
+            SessionPropertyManager sessionPropertyManager)
+
+    {
+        accessControl.checkCanSetUser(identity.getPrincipal().orElse(null), identity.getUser());
+
+        SessionBuilder sessionBuilder = Session.builder(sessionPropertyManager)
+                .setQueryId(queryId)
+                .setIdentity(identity)
```

Comment:
[Note]
It will leak Identity between Sessions if we ever decide to make it mutable. This seems like a bad idea especially that java has no mechanism of forcing immutability and Identity name doesn't suggest that it is immutable by convention.